[DNM][requires-evolution] Explicit varargs expansion expression

owenv opened the pull request:
This is a WIP prototype of a new expression, #variadic, which can be used to pass array elements as variadic arguments. I should have a post with the corresponding proposal pitch up on the forums shortly.

This isn’t meant to be a finalized implementation, just enough to demo the pitched feature and collect feedback. It’s been working well for me so far and passing tests, but still needs:

- Better test coverage
- Better compiler fixits in a few areas
- Some refactoring of argument coercion
- General cleanup and refactoring

owenv committed:
Improve the varargs conversion diag

owenv committed:
[Parse] Add initial support for parsing a new #variadic directive as a VarargExpansionExpr

owenv committed:
[Sema] Tiny CSGen changes needed to support explicit VarargExpansionExprs

owenv committed:
Clean up #variadic parsing

owenv committed:
Add new fixit offering to wrap arrays with #variadic

owenv committed:
better diags for improper #variadic usage

owenv committed:
Miscellaneous bug fixes and tests for #variadic

hamishknight commented on lib/Sema/CSDiag.cpp:
```diff
@@ -2296,6 +2297,34 @@ bool FailureDiagnosis::diagnoseContextualConversionError(
   }
 
+  // Diagnose passing '[T]' instead of a sequence of 'T' to a variadic argument
+  // of type 'T...'.
```
It would be nice if we could implement this using the new diagnostics framework (see CSFix & CSDiagnostics), as we're currently trying to move stuff out of CSDiag.cpp.

What you'll want to do is add a new constraint fix to CSFix.h, and then add it as a viable fix within `repairFailures` when the locator's last element is an `ApplyArgToParam` and you're dealing with an array type being passed to a variadic param. Unfortunately I don't believe there's currently an easy way of accessing the parameter flags from `repairFailures`, the simplest solution would probably be to store them on the constraint locator – see https://github.com/hamishknight/swift/commit/62d97fc39dda86a1145fd4c93096be10fc30c9f7 for an example of that.

Then you'll want to add a case to `simplifyFixConstraint` where you try to match the array element types and record the fix if successful.

owenv replied:
@hamishknight sounds good, I'll look into doing that. Thanks for the example!

hamishknight reviewed (commented):
Great to see this being implemented!

owenv committed:
Port variadic conversion diagnostics to the new diags framework.

This implementation isn't quite right yet because we're not yet checking the parameter flags to see if the param is actually variadic.

owenv committed:
Store the ParameterTypeFlags on ApplyArgToParam constraint locators

Cherry-picks https://github.com/hamishknight/swift/commit/7e3a6015556c6a3b6de3a344579e6d14ec3765f6 with some deletions
This is needed to support the ExpandArrayIntoVarargs ConstraintFix, which needs to be able to determine whether a parameter is variadic

owenv committed:
Remove original CSDiag.cpp implementation of variadic conversion diags

owenv commented:
This still needs some work, but if anyone with CI privileges is following along, mind starting a test run w/ the standard & validation suites? I see a lot of failures in the validation suite right now that read `no such module 'StdlibCollectionUnittest'` and I'm not sure if it's something I broke or an issue with my build setup.

hamishknight commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -2287,6 +2279,15 @@ bool ConstraintSystem::repairFailures(
+
+    auto elementTy = lhs->getArrayElementType();
+    if (elt.getKind() == ConstraintLocator::ApplyArgToParam &&
+        elementTy && elt.getParameterFlags().isVariadic()) {
```
You could use `ConstraintSystem::isArrayType` here instead, which saves you from having to introduce `isArrayDecl` & `getArrayElementType`.

hamishknight replied:
Also you no longer need to check for `elt.getKind() == ConstraintLocator::ApplyArgToParam` now, as there's an early exit for all other cases a few lines above.

hamishknight commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -1263,6 +1263,14 @@ ERROR(expr_selector_expected_property_expr,PointsToFirstBadToken,
+ERROR(expr_varargs_expansion_expected_expr,PointsToFirstBadToken,
+     "expected expression within '#variadic(...)'", ())
+ERROR(expr_varargs_expansion_expected_rparen,PointsToFirstBadToken,
+     "expected ')' to complete '#variadic' expression", ())
```
Nit: Strings should be indented by one more character to align with `expr_[...]`.

hamishknight commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -356,6 +356,23 @@ ERROR(cannot_convert_default_arg_value_nil,none,
+      "#variadic cannot be used alongside additional variadic arguments",
+      ())
+ERROR(pound_variadic_outside_arg_position,none,
+     "#variadic can only be used in an argument position",
```
Same here

hamishknight commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -356,6 +356,23 @@ ERROR(cannot_convert_default_arg_value_nil,none,
+     "remove brackets to pass array elements directly", ())
+NOTE(suggest_pass_elements_using_pound_variadic,none,
+    "wrap the expression in '#variadic(...)' to pass array elements as "
+    "variadic arguments", ())
```
And here

hamishknight commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -2287,6 +2279,15 @@ bool ConstraintSystem::repairFailures(
+    if (elt.getKind() == ConstraintLocator::ApplyArgToParam &&
+        elementTy && elt.getParameterFlags().isVariadic()) {
+      conversionsOrFixes.push_back(ExpandArrayIntoVarargs::create(
+          *this, lhs, rhs, getConstraintLocator(locator)));
```
Nit: You can use `loc` instead of `getConstraintLocator(locator)`.

hamishknight commented on test/Constraints/diagnostics.swift:
```diff
@@ -300,6 +300,59 @@ func rdar21784170() {
+variadic(arrayWithOtherEltType) // expected-error {{cannot convert value of type '[String]' to expected argument type 'Int'}}
+variadic(1, arrayWithOtherEltType) // expected-error {{cannot convert value of type '[String]' to expected argument type 'Int'}}
+variadic(["hello", "world"]) // expected-error 2 {{cannot convert value of type 'String' to expected element type 'Int'}}
+// expected-error@-1 {{cannot pass an array of type '[Int]' as variadic arguments of type 'Int'}}
```
Hmm, that's unfortunate. The good news at least is that it's not a specific problem with the fix you've introduced, it's a more general problem when we have multiple contextual fixes in play. I've filed a bug to track that: https://bugs.swift.org/browse/SR-11104.

owenv replied:
Thanks for taking the time to file this!

hamishknight reviewed (commented):
Thanks for working on those changes! I just have a few minor comments, I'll leave the rest for the folks at Apple.

owenv committed:
Post-rebase refactoring of #variadic

owenv committed:
Initial implementation of the new `as T...` syntax for array-to-varargs coercion

- The diags and tests have not yet been updated to reflect the new syntax
- The old syntax hasn't been removed yet
- But, some type checking hacks the old syntax used have been removed.

owenv committed:
Remove #variadic syntax in favor of 'as T...'

owenv committed:
Update interpreter test to use the new as T... syntax for varargs expansion

owenv committed:
working on tests and fixes for as T...

owenv committed:
fix compiler crash when trying to coerce an array to varargs with a non-variadic argument

owenv committed:
Reenable diagnostic requiring as T... only appears in an argument position

owenv committed:
Remove old syntax support for the abandoned #variadic syntax

owenv committed:
Fixup most of the as T... diags to reflect the new syntax

owenv committed:
Conversion diag fixes for as T...

owenv committed:
Remove the last few references to the #variadic syntax for vararg expansion

owenv committed:
more varargs expansion refactoring

hamishknight commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -118,7 +118,15 @@ ParserResult<Expr> Parser::parseExprAs() {
-    parsed = new (Context) CoerceExpr(asLoc, type.get());
+    bool includesVarargExpansion = false;
+    auto typeRepr = type.get();
+    if (Tok.isEllipsis()) {
```
You'll want to keep the `Tok.is(tok::oper_postfix)` check here to make sure we reject things like:

```swift
func foo(_ x: Int...) {}
foo([0] as Int ...)
```

In addition, it'll allow us to continue accepting the following:

```swift
let r = 0 as Int8...5
```

which calls the infix `...` operator.

owenv replied:
🤦‍♂ Good catch, thanks. Going to work on adding some more parsing tests

owenv committed:
Fix as T... ellipsis parsing bug, add a couple tests to make sure we maintain source compatibility with the infix ... operattor

owenv committed:
Better parser recovery in the event someone tries to use as! or as? for vararg expansion

owenv committed:
Have VarargExpansionExpr forward source locations to its subexpr again

owenv committed:
Add a couple more vararg expansion tests

owenv commented:
Closing this since it's pretty hopelessly out of date and the diagnostics improvements have landed elsewhere.
